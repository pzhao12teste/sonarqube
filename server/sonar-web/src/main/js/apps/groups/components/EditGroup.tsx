/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Form from './Form';
import { Group } from '../../../app/types';
import { translate } from '../../../helpers/l10n';
import { omitNil } from '../../../helpers/request';

interface Props {
  children: (props: { onClick: () => void }) => React.ReactNode;
  group: Group;
  onEdit: (data: { description?: string; id: number; name?: string }) => Promise<void>;
}

interface State {
  modal: boolean;
}

export default class EditGroup extends React.PureComponent<Props, State> {
  mounted: boolean;
  state: State = { modal: false };

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleClick = () => {
    this.setState({ modal: true });
  };

  handleClose = () => {
    if (this.mounted) {
      this.setState({ modal: false });
    }
  };

  handleSubmit = ({ name, description }: { name: string; description: string }) => {
    const { group } = this.props;
    return this.props.onEdit({
      description,
      id: group.id,
      // pass `name` only if it has changed, otherwise the WS fails
      ...omitNil({ name: name !== group.name ? name : undefined })
    });
  };

  render() {
    return (
      <>
        {this.props.children({ onClick: this.handleClick })}
        {this.state.modal && (
          <Form
            confirmButtonText={translate('update_verb')}
            group={this.props.group}
            header={translate('groups.update_group')}
            onClose={this.handleClose}
            onSubmit={this.handleSubmit}
          />
        )}
      </>
    );
  }
}
