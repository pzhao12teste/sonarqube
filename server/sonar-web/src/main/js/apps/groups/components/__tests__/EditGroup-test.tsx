/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import EditGroup from '../EditGroup';

it('should edit group', () => {
  const group = { id: 3, name: 'Foo', membersCount: 5 };
  const onEdit = jest.fn();
  const newDescription = 'bla bla';
  let onClick: any;

  const wrapper = shallow(
    <EditGroup group={group} onEdit={onEdit}>
      {props => {
        ({ onClick } = props);
        return <button />;
      }}
    </EditGroup>
  );
  expect(wrapper).toMatchSnapshot();

  onClick();
  wrapper.update();
  expect(wrapper).toMatchSnapshot();

  // change name
  wrapper.find('Form').prop<Function>('onSubmit')({ name: 'Bar', description: newDescription });
  expect(onEdit).lastCalledWith({ description: newDescription, id: 3, name: 'Bar' });

  // change description
  wrapper.find('Form').prop<Function>('onSubmit')({
    name: group.name,
    description: newDescription
  });
  expect(onEdit).lastCalledWith({ description: newDescription, id: group.id });

  wrapper.find('Form').prop<Function>('onClose')();
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});
