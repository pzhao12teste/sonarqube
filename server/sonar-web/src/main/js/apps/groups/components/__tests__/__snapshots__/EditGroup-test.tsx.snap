// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should edit group 1`] = `
<React.Fragment>
  <button />
</React.Fragment>
`;

exports[`should edit group 2`] = `
<React.Fragment>
  <button />
  <Form
    confirmButtonText="update_verb"
    group={
      Object {
        "id": 3,
        "membersCount": 5,
        "name": "Foo",
      }
    }
    header="groups.update_group"
    onClose={[Function]}
    onSubmit={[Function]}
  />
</React.Fragment>
`;

exports[`should edit group 3`] = `
<React.Fragment>
  <button />
</React.Fragment>
`;
